### ISSUE TITLE
Deepstream Custom Binding Behavior / Binding Mode / Binding Language

### ISSUE THREAD

**Issue by Svenskunganka** [aurelia/binding#310]
### Intro
I'm currently in the process of building a plugin for Aurelia that adds custom binding behaviors / binding modse / binding languages with [deepstream.io-client](https://github.com/hoxton-one/deepstream.io-client-js) which in simple terms enables simple, developer-friendly *three-way* databinding between Aurelia and a Deepstream Server.

I've been discussing this a bit with @jdanyow (got permission to post an issue [here](https://gitter.im/Aurelia/Discuss?at=56b378f365d6cbe729041cf0)) on how I would go about implementing this the best way and I'm opening this issue to start an open discussion between community members & the Aurelia team.
I also need some guidance as I don't have much experience or knowledge with the internals of Aurelia and some information regarding the conventions I should follow.

### Ideas & Examples

Some ideas I've had is to either add Binding Behaviors, Binding Modes or a Binding Language. The idea is to *extend* Aurelias current bindings, e.g `value.bind`, `click.delegate`, etc. I've looked through the source of every library that is about binding or templating and it looks like creating a Binding Mode is out of the question as it doesn't look extendable. A Binding Language seems to *override* Aurelias current Binding Language when I instead want to *extend* it, but please correct me if I'm wrong.

**1. Bi-directional three-way data binding**
With Deepstream, you can subscribe to changes that happens in a `record` (table/document/etc) on *property level*, but you can also subscribe to the entire record. Below is an example of how you would do it with standalone deepstream client:
```javascript
var mike = client.record.getRecord('users/mike') // 'mike' in this case is a UID

var favoriteColorChanged = function(color) {...}

mike.subscribe('favoriteColor', favoriteColorChanged)
```
The above snippets `favoriteColorChanged` function will be called whenever the `favoriteColor` property in the record changes.

Ideally, I'd like to implement this in Aurelia's Binding Modes so I can do this:
```html
<input type="text" value.stream="users.mike.favoriteColor">
// or
<p>${users.mike.favoriteColor & stream}</p>
// another that raises an issue
<p>${users[navigationInstruction.params.uid].favoriteColor}</p>
```
The above snippet would then update the record whenever the value of the input changes, but also update the value whenever the record changes. The issue with the above snippet is that I'd have to parse `users.mike.favoriteColor` to a valid deepstream path: `users/mike.favoriteColor` and that I'm not sure that I am following the convetions of Aurelia.

Some other implementations of this:
```html
<input type="text" value.bind="'users/mike.favoriteColor' & stream">
// or
<input type="text" value.bind="path & stream"> // same as above, but `path` is a property on the VM that holds the same string as the example above
```
These implementations makes it a bit easier for me to implement as it only requires a custom Binding Behavior, but raises an issue as the `value` attribute is not bound to the `path`, but rather to the values Deepstream returns.

I'm not sure what the best implementation is, which is why I'm asking both the community and the Aurelia team.

**2. Calling RPC's directly from the DOM**
Deepstream also has support for RPC (Remote Procedure Calls). Clients can both call and register themselves for RPC's and when an RPC is called, deepstream server will evenly distribute the calls to the registered RPC listeners.

An example of calling an RPC with deepstream client:
```javascript
client.rpc.make('sendPasswordResetEmail', email, function(err, res) {
  // Handle response
})
```
An example of the same implementation in Aurelia:
```html
<button click.rpc="sendPasswordResetEmail(email) & call:myFunc">Send</button>
// and
<button click.rpc="sendPasswordResetEmail(email) & set:myProperty">Send</button>
```
The above example raises many issues as it requires custom parsing that overrides Aurelias default parsing mechanism.
The `call:myFunc` in this case *calls* a method on the VM with the response of the RPC.
The `set:myProperty` in this case *sets* a property on the VM to the response of the RPC.

Other implementations:
```html
<button click.trigger="'sendPasswordResetEmail' & param:email & call:myFunc">Send</button>
```

**3. Repeatable Events**
Deepstream also have events.

Deepstream examples:
```javascript
// Subscribing to events:
client.event.subscribe('chat', function(res) {
  // Handle response...
})

// Emitting events:
client.event.emit('chat', 'Hello World')
```

Aurelia implementation:

Repeat over event:
```html
<p repeat.event="message of chat">${message}</p>
```
Other implementations:
```html
<p repeat.for="message of 'chat' & event">${message}</p>
```

Emit event:
```html
<button click.event="chat('Hello World')">Send</button>
```
Other implementations:
```html
<button click.trigger="'Hello World' & event:'chat'">Send</button>
```

### Finishing up
Some of these implementations are pretty hacky and I'd love to get some insight on what is possible to do with Aurelia without *overriding* everything, but rather *extending* it. If you have some good feedback on other implementations they are very welcome.

cc @jdanyow @EisenbergEffect

**Comment by jdanyow** [comment]
Hi @Svenskunganka - here's some info on extending the binding language:

`bind`, `trigger`, `delegate`, `one-way`, `for`, etc are what's known as "binding commands".  These commands are implemented in methods on the SyntaxInterpreter class. You can add your own binding commands with the following steps:

```shell
jspm install aurelia-templating-binding
```

**main.js**
```js
import {SyntaxInterpreter} from 'aurelia-templating-binding';

SyntaxInterpreter.foo = function(resources, element, info, existingInstruction) {
  // note: "this" will be the SyntaxInterpreter instance (you'll have access to the SyntaxInterpreter's members via this)
  //   ... see examples in SyntaxInterpreter class ...
}
```
https://github.com/aurelia/templating-binding/blob/master/src/syntax-interpreter.js#L184

Some thoughts on the scenarios you've described:

1. Bi-directional three-way data binding
 This seems like the core/primary scenario, and would most likely involve a custom binding command (see above)
2. Calling RPC's directly from the DOM
 Recommend doing something like `<button click.rpc="sendPasswordResetEmail(email)" click.trigger="myFunc()">Send</button>` instead of attempting to do both in the same command with `<button click.rpc="sendPasswordResetEmail(email) & call:myFunc">Send</button>`
I'm unclear how the binding command will get the correct `client` instance- I'm assuming there would be some configuration necessary and potentially multiple clients in the same app.
3. This is a lot like Rx Observables and should be consistent with whatever we do for that. On the other hand you could experiment with adding a repeat strategy that supports this scenario: https://github.com/aurelia/templating-resources/blob/master/src/repeat-strategy-locator.js#L39

**Comment by jdanyow** [comment]
Hi @Svenskunganka - how's this going? Need any help?

**Comment by Svenskunganka** [comment]
Hey @jdanyow, sorry I haven't been keeping contact.

I've had to put this on hold for a while, but it is still definitely going to happen. Have been busy with installing new server hardware in our datacenter, provisioning and migrating our old projects to our new infrastructure. Deepstream server also had some missing points that I had to adress before I could start development.

I'll poke you on Gitter if I need some clarifications.
